test_conformance/extensions/cl_khr_semaphore/test_semaphores_cross_queue.cpp

```diff
@@ -0,0 +1,321 @@
+        test_error(error, "Failed to create program with source");
+
+        error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+        test_error(error, "Failed to build program");
+
+        kernel = clCreateKernel(program, "copy", &error);
+        test_error(error, "Failed to create copy kernel");
+
+        // create producer/consumer out-of-order queues
+        producer_queue = clCreateCommandQueue(
+            context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &error);
+        test_error(error, "Could not create command queue");
+
+        if (single_queue)
+        {
+            consumer_queue = producer_queue;
+        }
+        else
+        {
+            consumer_queue = clCreateCommandQueue(
+                context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE,
+                &error);
+            test_error(error, "Could not create command queue");
+        }
+
+        // Create semaphore
+        cl_semaphore_properties_khr sema_props[] = {
+            static_cast<cl_semaphore_properties_khr>(CL_SEMAPHORE_TYPE_KHR),
+            static_cast<cl_semaphore_properties_khr>(
+                CL_SEMAPHORE_TYPE_BINARY_KHR),
+            0
+        };
+        semaphore =
+            clCreateSemaphoreWithPropertiesKHR(context, sema_props, &error);
+        test_error(error, "Could not create semaphore");
+
+        // create memory resources
+        in_mem = clCreateBuffer(context, CL_MEM_READ_ONLY,
+                                sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        out_mem = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                                 sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        error = clSetKernelArg(kernel, 0, sizeof(in_mem), &in_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(kernel, 1, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+        cl_int err = SetUp();
+        test_error(err, "SetUp failed");
+
+        clEventWrapper wait_events[2];
+        // enqueue producer operations and signal semaphore
+        const cl_int pattern_pri = 42;
+        const cl_int pattern_sec = 0xACDC;
+
+        std::vector<cl_int> host_buffer(num_elems, pattern_pri);
+
+        err = clEnqueueWriteBuffer(
+            producer_queue, in_mem, CL_FALSE, 0, sizeof(cl_int) * num_elems,
+            host_buffer.data(), 0, nullptr, &wait_events[0]);
+        test_error(err, "clEnqueueReadBuffer failed");
+
+        err = clEnqueueFillBuffer(producer_queue, in_mem, &pattern_sec,
+                                  sizeof(cl_int), 0, sizeof(cl_int) * num_elems,
+                                  1, &wait_events[0], nullptr);
+        test_error(err, "clEnqueueFillBuffer failed");
```

Comment:
Replaced with two producer + two consumer operations:

1) each producer operation writing half of input buffer simultaneously
2) each consumer operation copies half input to output buffer simultaneously (with modified kernel with offset argument)

side note: we could replace `clEnqueueBarrierWithWaitList` calls with wait-events - effect the same but slightly cleaner code.

Reply:
I'm sorry, I don't think I'm explaining this very well.  Catch me on Slack if you'd like and we can have a more interactive conversation.

By "multiple producers" I'm thinking of two commands in the same queue, not two separate producer queues.  I'm also thinking that each command should produce into a unique memory object, so not writing into the same memory memory object, which is probably a data race.

Here's the original snip from #1979 with a bit more detail what I am looking for.  I am going to use "fill" as my producer and "copy" as my consumer, but other options are also viable:

```c++
// Producers:
// Note, these commands should have no event dependencies and produce no events!
clEnqueueFillBuffer(producer_queue, temp_data_a, pattern_a);
clEnqueueFillBuffer(producer_queue, temp_data_b, pattern_b);

// This barrier requires all producer operations to be complete:
clEnqueueBarrierWithWaitList(producer_queue); 

// The semaphore cannot be signaled until the barrier is complete:
clEnqueueSignalSemaphore(producer_queue, semaphore);

// Note: the consumer_queue can be a separate queue or the same as the producer_queue!
// If the consumer_queue and the producer_queue are the same then there needs to be
// an explicit dependency between the semaphore signal and the semaphore wait, which
// could be an event dependency or another barrier.

// Consumers:
// Wait for the producers to signal completion via the semaphore:
clEnqueueWaitSemaphore(consumer_queue, semaphore);

// This barrier ensures all consumers cannot start until the semaphore wait is complete:
clEnqueueBarrierWithWaitList(consumer_queue);

// Note: these commands should also have no event dependencies and produce no events!
clEnqueueCopyBuffer(consumer_queue, temp_data_a, result_data_a);
clEnqueueCopyBuffer(consumer_queue, temp_data_b, result_data_b);

clEnqueueReadBuffer(consumer_queue, result_data_a, ...); // should read pattern_a
clEnqueueReadBuffer(consumer_queue, result_data_b, ...); // should read pattern_b
```

Reply:
Understood... I hope :). I reverted single producer/consumer queues and doubled input/output buffers. In order to copy buffers instead of `clEnqueueCopyBuffer` I am using previous kernels, please let me know if that works for you.